File: src/main/java/com/hjj/apiserver/repositroy/impl/PurchaseRepositoryImpl.java
package com.hjj.apiserver.repositroy.impl;

import com.hjj.apiserver.domain.purchase.Purchase;
import com.hjj.apiserver.dto.PurchaseDto;
import com.hjj.apiserver.repositroy.PurchaseRepositoryCustom;
import com.querydsl.jpa.impl.JPAQueryFactory;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Pageable;

import java.util.ArrayList;
import java.util.List;


@RequiredArgsConstructor
public class PurchaseRepositoryImpl implements PurchaseRepositoryCustom {

    private final JPAQueryFactory jpaQueryFactory;

    public PurchaseDto.ResponsePurchaseDetail findPurchase(Long userNo, Long purchaseNo){
//        return jpaQueryFactory
//                .select(new QPurchaseDto_ResponsePurchaseDetail(
//                        purchaseEntity.accountBookEntity.accountBookNo,
//                        purchaseEntity.cardEntity.cardNo,
//                        categoryEntity.parentCategory.categoryNo,
//                        purchaseEntity.categoryEntity.categoryNo,
//                        purchaseEntity.storeName,
//                        purchaseEntity.purchaseType,
//                        purchaseEntity.price,
//                        purchaseEntity.reason,
//                        purchaseEntity.purchaseDate
//                ))
//                .from(purchaseEntity)
//                .leftJoin(purchaseEntity.categoryEntity, categoryEntity)
//                .where(purchaseEntity.purchaseNo.eq(purchaseNo).and(purchaseEntity.userEntity.userNo.eq(userNo)))
//                .fetchOne();

        return null;

    }

    @Override
    public List<Purchase> findPurchasePageCustom(PurchaseDto purchaseDto, Pageable pageable) {
//        return jpaQueryFactory
//                .select(purchaseEntity)
//                .from(purchaseEntity)
//                .leftJoin(purchaseEntity.cardEntity, cardEntity).fetchJoin()
//                .leftJoin(purchaseEntity.categoryEntity, categoryEntity).fetchJoin()
//                .where(
//                        purchaseEntity.purchaseDate.between(purchaseDto.getStartDate(), purchaseDto.getEndDate())
//                                .and(purchaseEntity.accountBookEntity.accountBookNo.eq(purchaseDto.getAccountBookNo())
//                                        .and(purchaseEntity.deleteYn.eq('N'))))
//                .offset(pageable.getOffset())
//                .limit(pageable.getPageSize() + 1)
//                .orderBy(purchaseEntity.purchaseDate.desc(), purchaseEntity.purchaseNo.desc())
//                .fetch();
        return new ArrayList<>();
    }



}

File: src/main/kotlin/com/hjj/apiserver/repository/purchase/PurchaseRepositoryImpl.kt
package com.hjj.apiserver.repository.purchase

import com.hjj.apiserver.domain.card.QCard.Companion.card
import com.hjj.apiserver.domain.category.QCategory.Companion.category
import com.hjj.apiserver.domain.purchase.Purchase
import com.hjj.apiserver.domain.purchase.QPurchase.Companion.purchase
import com.hjj.apiserver.dto.purchase.response.PurchaseDetailResponse
import com.querydsl.core.types.Projections
import com.querydsl.jpa.impl.JPAQueryFactory
import org.springframework.data.domain.Pageable
import java.time.LocalDate

class PurchaseRepositoryImpl(
    private val jpaQueryFactory: JPAQueryFactory,
) : PurchaseRepositoryCustom {
    override fun findPurchase(
        userNo: Long,
        purchaseNo: Long,
    ): PurchaseDetailResponse? {
        return jpaQueryFactory
            .select(
                Projections.constructor(
                    PurchaseDetailResponse::class.java,
                    purchase.purchaseNo,
                    purchase.accountBook.accountBookNo,
                    purchase.card.cardNo,
                    category.categoryNo,
                    category.parentCategory.categoryNo,
                    purchase.purchaseType,
                    purchase.price,
                    purchase.reason,
                    purchase.purchaseDate,
                ),
            )
            .from(purchase)
            .leftJoin(purchase.category, category)
            .where(purchase.purchaseNo.eq(purchaseNo).and(purchase.userEntity.userNo.eq(userNo)))
            .fetchOne()
    }

    override fun findPurchasePageCustom(
        searchStartDate: LocalDate,
        searchEndDate: LocalDate,
        accountBookNo: Long,
        pageable: Pageable,
    ): List<Purchase> {
        return jpaQueryFactory
            .select(purchase)
            .from(purchase)
            .leftJoin(purchase.card, card).fetchJoin()
            .leftJoin(purchase.category, category).fetchJoin()
            .where(
                purchase.purchaseDate.between(searchStartDate, searchEndDate)
                    .and(purchase.accountBook.accountBookNo.eq(accountBookNo))
                    .and(purchase.isDelete.isFalse),
            )
            .offset(pageable.offset)
            .limit(pageable.pageSize + 1L)
            .orderBy(purchase.purchaseDate.desc(), purchase.purchaseNo.desc())
            .fetch()
    }
}
